src/dstack/_internal/server/services/logs.py

```diff
@@ -1,117 +1,342 @@
-            for line in f:
-                log_event = LogEvent.__response__.parse_raw(line)
-                if request.start_time and log_event.timestamp <= request.start_time:
-                    continue
-                if request.end_time is None or log_event.timestamp < request.end_time:
-                    logs.append(log_event)
-                else:
-                    break
-    except IOError:
-        pass
-    if request.descending:
-        logs = list(reversed(logs))
-    return JobSubmissionLogs(logs=logs)
-
-
-def _runner_log_event_to_log_event(runner_log_event: RunnerLogEvent) -> LogEvent:
-    return LogEvent(
-        timestamp=datetime.fromtimestamp(runner_log_event.timestamp / 1e9, tz=timezone.utc),
-        log_source=LogEventSource.STDOUT,
-        message=base64.b64encode(runner_log_event.message).decode(),
-    )
-
-
-def _get_job_log_file_path(project_name: str, run_name: str, job_submission_id: UUID) -> Path:
-    return (
-        settings.SERVER_DIR_PATH
-        / "projects"
-        / project_name
-        / "logs"
-        / run_name
-        / str(job_submission_id)
-        / "job.log"
-    )
-
-
-def _get_runner_log_file_path(project_name: str, run_name: str, job_submission_id: UUID) -> Path:
-    return (
-        settings.SERVER_DIR_PATH
-        / "projects"
-        / project_name
-        / "logs"
-        / run_name
-        / str(job_submission_id)
-        / "runner.log"
-    )
+
+    def _runner_log_event_to_log_event(self, runner_log_event: RunnerLogEvent) -> LogEvent:
+        return LogEvent(
+            timestamp=_unix_time_ms_to_datetime(runner_log_event.timestamp),
+            log_source=LogEventSource.STDOUT,
+            message=_b64encode_raw_message(runner_log_event.message),
+        )
+
+
+_default_log_storage: Optional[LogStorage] = None
+
+
+def get_default_log_storage() -> LogStorage:
+    global _default_log_storage
+    if _default_log_storage is not None:
+        return _default_log_storage
+    if settings.SERVER_CLOUDWATCH_LOG_GROUP:
+        if BOTO_AVAILABLE:
+            try:
+                _default_log_storage = CloudWatchLogStorage(
+                    group=settings.SERVER_CLOUDWATCH_LOG_GROUP,
+                    region=settings.SERVER_CLOUDWATCH_LOG_REGION,
+                )
+            except LogStorageError as e:
+                logger.warning("Failed to initialize CloudWatch Logs storage: %s", e)
+            else:
+                logger.debug("Using CloudWatch Logs storage")
+        else:
+            logger.warning("Cannot use CloudWatch Logs storage, boto3 is not installed")
+    if _default_log_storage is None:
+        logger.debug("Using file-based storage")
+        _default_log_storage = FileLogStorage()
+    atexit.register(_default_log_storage.close)
+    return _default_log_storage
+
+
+_default_log_storage = get_default_log_storage()
+
+
+poll_logs = _default_log_storage.poll_logs
+write_logs = _default_log_storage.write_logs
```

Comment:
The functions are synchronous and they are called synchronously too. The calls should be made async (e.g. wrapped with `run_async`) to avoid blocking the event loop.

Reply:
This service is used in both sync and async contexts (more specifically, `write_logs` is used in a sync function already executed inside an executor, and `poll_logs` is used in an async FastAPI handler). I've added both sync and async global proxy functions for the default log storage methods: https://github.com/dstackai/dstack/pull/1597/commits/2df00938cc2e174d33caf5b57360f96d80f2625f